// app-configuration/unauthorizedAppConfigurationClient.ts
/**
 * Mock App config Client, that throws an 401 status code 
*/
import { ConfigurationSetting } from "@azure/app-configuration";
import { RestError, RawHttpHeaders, PipelineRequest, createHttpHeaders, HttpHeaders } from "@azure/core-rest-pipeline";

class UnauthorizedAppConfigurationClient {

    public async listConfigurationSettings(): Promise<IteratorResult<ConfigurationSetting, any>> {
        const rawHttpHeaders: RawHttpHeaders = {
            "user-agent": "AzurePipelines.AzureAppConfiguration.Push",
            "x-ms-client-request-id": "12345908",
            "www-authenticate":"HMAC"
        };
        
        const httpHeaders: HttpHeaders = createHttpHeaders(rawHttpHeaders);
        const request: PipelineRequest = {
            url: "https://test.azconfig.io",
            method: "GET", 
            headers: httpHeaders,
            timeout: 1000,
            withCredentials: false,
            requestId: "request-id"
        }

        return Promise.reject(new RestError('', { statusCode: 401, 
            request: request,
            response: {
                headers: httpHeaders,
                status: 401,
                request: request
            }
        }));
    }
}

exports.AppConfigurationClient = UnauthorizedAppConfigurationClient;
